### ISSUE TITLE
Create a "diagnose" command

### ISSUE THREAD

**Issue by mattstauffer** [laravel/valet#936]
@drbyte regularly works hard to get useful output from folks as he's doing free customer support in the issues here.

I thought it might be nice if we could create a command or two that makes his job easier.

The first, I think, would be "diagnose", which gives him the output he always asks for on issues.

We could then create an issue template that requires the output of "diagnose".

I think the second one, later, would be more work, but could be more like a "doctor" command where it tries to actually find out what's wrong and possibly even tries to fix it. But right now, I think, it would just give information about the system.

Here's the commands he asked the last person I saw this with to run:

```
sw_vers
valet --version
cat ~/.config/valet/config.json
cat ~/.composer/composer.json
ls -al /etc/sudoers.d/
brew config
brew services list
brew list --versions | grep -E '(php|nginx|dnsmasq|mariadb|mysql|mailhog|openssl)(@\d\..*)?\s'
php -v
which -a php
php --ini
nginx -v
which -a php-fpm
sudo /usr/local/opt/php/sbin/php-fpm --test
sudo /usr/local/opt/php/sbin/php-fpm -y /usr/local/etc/php/7.4/php-fpm.conf --test
composer global diagnose
```

The sudo items are the only concern I have, so we might have to see whether those are optional or something.

**Comment by drbyte** [comment]
Here's a longer list of things I've collected. I don't always ask for all of these if the context isn't related.

```sh
sw_vers
valet --version
cat ~/.config/valet/config.json
cat ~/.composer/composer.json
composer global diagnose
ls -al /etc/sudoers.d/
brew config
brew services list
brew list --versions | grep -E '(php|nginx|dnsmasq|mariadb|mysql|mailhog|openssl)(@\d\..*)?\s'
php -v
which -a php
php --ini
nginx -v
curl --version
php --ri curl
~/.composer/vendor/laravel/valet/bin/ngrok version
ls -al ~/.ngrok2
brew info nginx
brew info php
brew info openssl
openssl version -a
openssl ciphers
sudo nginx -t
which -a php-fpm
sudo /usr/local/opt/php/sbin/php-fpm --test
sudo /usr/local/opt/php/sbin/php-fpm -y /usr/local/etc/php/7.4/php-fpm.conf --test
ls -al ~/Library/LaunchAgents
ls -al /Library/LaunchAgents
ls -al /Library/LaunchDaemons
```

Also, try the following:
```sh
brew services cleanup
brew cleanup
brew doctor
composer global update
composer self-update
```

**Comment by mattstauffer** [comment]
So what if we basically did it so each of those commands run, and then we copied it.. and maybe a later version would actually generate the markdown enclosing each of those with a closed section (using this: https://twitter.com/stauffermatt/status/1253001825327951872) so you could just expand only the sections you want?

Thiinking it might push that content into the clipboard using pbcopy...

**Comment by mattstauffer** [comment]
OOh, also:

If anything's not up to date it can prompt them to update them first.

E.g. like @drbyte did here:

https://github.com/laravel/valet/issues/928#issuecomment-618449793

**Comment by aryehraber** [comment]
Now that the `diagnose` command is (almost) ready for prime time, would it be worth discussing the `doctor` command to help guide the user through the most common issues/scenarios? 

@drbyte's mentioned these commands above:

```
brew services cleanup
brew cleanup
brew doctor
composer global update
composer self-update
```

Happy to help make this a reality if this hasn't been started yet!

**Comment by drbyte** [comment]
@aryehraber I just merged the latest master into my working branch where I'd been brainstorming ideas for a `doctor` command: https://github.com/drbyte/valet/commit/78807452b97191292b10513163a93059a064dfe1

Half of the code there is "reading" many of the `diagnose` commands we've discussed. There's more to be done with the results of several of them.

FYI - Some of the comments I left in are real to-do's, but some of them are just dreaming about nice-to-have features that are out-of-scope for the `doctor` command, such as supporting multiple users and multiple php versions simultaneously. I should remove them, but I figure there's no harm letting them be seen in case they inspire other PRs.

`doctor` should focus on identifying specific issues, fixing them if possible (whether safe to "just do" automatically, or if should prompt user first), or providing guidance on how to fix them if must be done manually.

I suppose one thing that could be done is split out some of the `diagnose` commands into their own domain-specific methods. I've done that a bit as `lint()` methods, but they're more generic than specific. If we add several "linters" in each domain that we can use for reading output, they could be shared for both the Doctor and Diagnose commands.

**Comment by drbyte** [comment]
Updated commit link: https://github.com/drbyte/valet/commit/ab130b47a1e70a039f1b77c886e29470860ca84e

**Comment by aryehraber** [comment]
@drbyte Nice one, will peruse through when I next have some spare time 👍

**Comment by drbyte** [comment]
This was closed by #938
